src/tools/pbs_upgrade_job.c

```diff
@@ -234,102 +311,284 @@ check_job_file(int fd)
+
+	return jf_19_20;
+}
+
+/**
+ * @brief	Upgrade 19-20 jobfix structure to latest
+ *
+ * @param[in]	old_jf - 19-20 jobfix struct
+ *
+ * @return	jobfix_19_20
+ * @retval	v19 converted jobfix
+ */
+struct jobfix
+convert_19jf_to_21(jobfix_19_20 old_jf)
+{
+	struct jobfix jf;
+
+	memset(&jf, 0, sizeof(jf));
+	jf.ji_jsversion = JSVERSION;
+	jf.ji_svrflags = old_jf.ji_svrflags;
+	jf.ji_stime = old_jf.ji_stime;
+	snprintf(jf.ji_jobid, sizeof(jf.ji_jobid), "%s", old_jf.ji_jobid);
+	snprintf(jf.ji_fileprefix, sizeof(jf.ji_fileprefix), "%s", old_jf.ji_fileprefix);
+	snprintf(jf.ji_queue, sizeof(jf.ji_queue), "%s", old_jf.ji_queue);
+	snprintf(jf.ji_destin, sizeof(jf.ji_destin), "%s", old_jf.ji_destin);
+	jf.ji_un_type = old_jf.ji_un_type;
+	memcpy(&jf.ji_un, &old_jf.ji_un, sizeof(jf.ji_un));
+
+	return jf;
+}
+
+union jobextend
+convert_19ext_to_21(union jobextend_19_20 old_extend)
+{
+	union jobextend je;
+
+	memset(&je, 0, sizeof(je));
+	snprintf(je.fill, sizeof(je.fill), "%s", old_extend.fill);
+	snprintf(je.ji_ext.ji_jid, sizeof(je.ji_ext.ji_jid), "%s", old_extend.ji_ext.ji_4jid);
+	je.ji_ext.ji_credtype = old_extend.ji_ext.ji_credtype;
+#ifdef PBS_MOM
+	je.ji_ext.ji_nodeidx = old_extend.ji_ext.ji_nodeidx;
+	je.ji_ext.ji_taskidx = old_extend.ji_ext.ji_taskidx;
+#if MOM_ALPS
+	je.ji_ext.ji_reservation = old_extend.ji_ext.ji_reservation;
+	je.ji_ext.ji_pagg = old_extend.ji_ext.ji_pagg;
+#endif
+#endif
+
+	return je;
+}
+
 /**
  * @brief
  *		Upgrade a job file from an earlier version.
  *
  * @param[in]	fd		-	File descriptor from which to read
+ * @param[in]	ver		-	Old version
  *
  * @return	int
- * @retval	-1	: failure
+ * @retval	1	: failure
  * @retval	 0	: success
  */
 int
-upgrade_job_file(int fd)
+upgrade_job_file(int fd, int ver)
 {
 	FILE *tmp = NULL;
 	int tmpfd = -1;
-	int len;
+	jobfix_19_20 qs_19_20;
+	jobfix_PRE19 old_jobfix_pre19;
+	union jobextend_19_20 old_ji_extended;
 	int ret;
 	off_t pos;
-	jobfix_PRE19 old_jobfix_pre19;
 	job new_job;
 	errno = 0;
+	int len;
+	svrattrl *pal = NULL;
+	char statechar;
+	svrattrl *pali;
+	svrattrl dummy;
+	char statebuf[2];
+	char ssbuf[5];
+	char *charstrm;
 
 	/* The following code has been modeled after job_recov_fs() */
 
-	/* Read in the pre19 jobfix structure */
-	memset(&old_jobfix_pre19, 0, sizeof(old_jobfix_pre19));
-	len = read(fd, (char *)&old_jobfix_pre19, sizeof(old_jobfix_pre19));
-	if (len < 0) {
-		fprintf(stderr, "Failed to read input file [%s]\n",
+	if (ver == 18) {
+		/* Read in the pre19 jobfix structure */
+		memset(&old_jobfix_pre19, 0, sizeof(old_jobfix_pre19));
+		len = read(fd, (char *) &old_jobfix_pre19, sizeof(old_jobfix_pre19));
+		if (len < 0) {
+			fprintf(stderr, "Failed to read input file [%s]\n",
 				errno ? strerror(errno) : "No error");
+			return 1;
+		}
+		if (len != sizeof(old_jobfix_pre19)) {
+			fprintf(stderr, "Format not recognized, not enough fixed data.\n");
+			return 1;
+		}
+
+		qs_19_20 = convert_pre18jf_to_19(old_jobfix_pre19);
```

Comment:
I'd change the function name to convert_pre19jf_to_19. It is currently confusing because the version read is not pre18.